Casting owned type to raw pointer has unclear error message

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.

If you cannot produce a minimal reproduction case (something that would work in
isolation), please provide the steps or even link to a repository that causes
the problematic output to occur. 
-->

Given the following code: <!-- Please provide a link to play.rust-lang.org -->

```rust
struct Something {
    pub field: u32,
}

fn main() {
    let something = Something {
        field: 1337,
    };
    
    let pointer_to_something = something as *const Something;
}
```
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1a1c55703dee4ed2078b5c704bd1f379

The current output is:

```
error[E0605]: non-primitive cast: `Something` as `*const Something`
  --> src/main.rs:10:41
   |
10 |     let pointer_to_something = something as *const Something;
   |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ an `as` expression can only be used to convert between primitive types or to coerce to a specific trait object
```

I think, that in the context of casting to a pointer, the output in this case should make it clear that the error could be fixed by taking a reference to the type and cast the reference to a raw pointer.  
Example:

```
error[E0605]: non-primitive cast: `Something` as `*const Something`
  --> src/main.rs:10:41
   |
10 |     let pointer_to_something = something as *const Something;
   |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot cast an owned type to a raw pointer
   |     consider borrowing here: &something as *const Something
```

<!--
If the problem is not self-explanatory, please provide a rationale for the
change.
-->

<!--
If dramatically different output is caused by small changes, consider also
adding them here.

If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions. The output might also be
different depending on the Edition.
-->

--- a/compiler/rustc_typeck/src/check/cast.rs
+++ b/compiler/rustc_typeck/src/check/cast.rs
@@ -359,6 +359,21 @@ impl<'a, 'tcx> CastCheck<'tcx> {
                     {
                         sugg = Some(format!("&{}", mutbl.prefix_str()));
                     }
+                } else if let ty::RawPtr(TypeAndMut { mutbl, .. }) = *self.cast_ty.kind() {
+                    if fcx
+                        .try_coerce(
+                            self.expr,
+                            fcx.tcx.mk_ref(
+                                &ty::RegionKind::ReErased,
+                                TypeAndMut { ty: self.expr_ty, mutbl },
+                            ),
+                            self.cast_ty,
+                            AllowTwoPhase::No,
+                        )
+                        .is_ok()
+                    {
+                        sugg = Some(format!("&{}", mutbl.prefix_str()));
+                    }
                 }
                 if let Some(sugg) = sugg {
                     err.span_label(self.span, "invalid cast");

--- a/src/test/ui/cast/issue-84213.fixed
+++ b/src/test/ui/cast/issue-84213.fixed
@@ -1,0 +1,15 @@
+// run-rustfix
+
+struct Something {
+    pub field: u32,
+}
+
+fn main() {
+    let mut something = Something { field: 1337 };
+
+    let _pointer_to_something = &something as *const Something;
+    //~^ ERROR: non-primitive cast
+
+    let _mut_pointer_to_something = &mut something as *mut Something;
+    //~^ ERROR: non-primitive cast
+}

--- a/src/test/ui/cast/issue-84213.rs
+++ b/src/test/ui/cast/issue-84213.rs
@@ -1,0 +1,15 @@
+// run-rustfix
+
+struct Something {
+    pub field: u32,
+}
+
+fn main() {
+    let mut something = Something { field: 1337 };
+
+    let _pointer_to_something = something as *const Something;
+    //~^ ERROR: non-primitive cast
+
+    let _mut_pointer_to_something = something as *mut Something;
+    //~^ ERROR: non-primitive cast
+}

--- a/src/test/ui/cast/issue-84213.stderr
+++ b/src/test/ui/cast/issue-84213.stderr
@@ -1,0 +1,25 @@
+error[E0605]: non-primitive cast: `Something` as `*const Something`
+  --> $DIR/issue-84213.rs:10:33
+   |
+LL |     let _pointer_to_something = something as *const Something;
+   |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ invalid cast
+   |
+help: borrow the value for the cast to be valid
+   |
+LL |     let _pointer_to_something = &something as *const Something;
+   |                                 ^
+
+error[E0605]: non-primitive cast: `Something` as `*mut Something`
+  --> $DIR/issue-84213.rs:13:37
+   |
+LL |     let _mut_pointer_to_something = something as *mut Something;
+   |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ invalid cast
+   |
+help: borrow the value for the cast to be valid
+   |
+LL |     let _mut_pointer_to_something = &mut something as *mut Something;
+   |                                     ^^^^
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0605`.

